darwin.stdenv: Update to LLVM 15 and clang to 15.0.7

reckenrode opened the pull request:
###### Description of changes

Bumps LLVM and clang to 15.0.7 on Darwin. This was prompted by #229210 because the 13.3 SDK does not build with LLVM 11. This PR reworks the Darwin stdenv to decouple it from the bootstrap tools is written not to need updates to bootstrap tools and includes changes to allow LLVM to be bumped without requiring a bump to the bootstrap tools.

See the comments in 309dc9d8f1e explaining the rework of the Darwin stden bootstrap process. My hope is this will make future bumps easier and more approachable for non-Darwin developers. It should also make experimentation a little safer. I’d like to see if any space could be saved by enabling LTO during the bootstrap, but that’s out of scope for this PR.

###### Issues to be resolved

These have all been resolved. The changes needed for the bootstrap are included in this PR. The others are being handled in separate PRs, which will be linked below.

- [X] Disable LLVM tests when building the stdenv. This causes an infinite recursion currently. I have a workaround, but ideally it should be disabled in the bootstrap.
- [X] `pkg-config` requires disabling a warning, or the build fails. It’s vendoring a very old version of GLib. Unfortunately, nixpkgs does not provide a way to use the nixpkgs `glib` package. I also tried, and it caused an infinite recursion.
- [X] `codesign` in bootstrap tools is too old for LLVM, but I’ve included a workaround. A bootstrap tools update would obviate the change.
- [X] Updating platform definition now that LLVM understands `apple-m1`. This should likely go in another PR, but it’s included here for testing. See #166401.

###### Testing Done

Running `nixpkgs-review` on a stdenv change is obviously going to rebuild an infeasible amount of packages. I have attempted to build and run a number of packages to catch common failures. I’ve tried to change things needed to build the stdenv from those that fix common problems. See below for PRs related to those (and sandbox issues I encountered along the way).

###### Related PRs

These fail to build due to the changes.

- #234117
- #234219

These are Darwin sandbox-related fixes. They’re not strictly necessary, but they allow Darwin users to build more things with the sandbox enabled.

- #234106
- #234111
- #234115
- #234119
- #234122
- #234125
- #234129
- #234132

This fix is related to running the store on case-sensitive APFS. It’s the first thing that has broken for me. Most case sensitivity issues during builds are due to `/tmp` being insensitive.

- #234105
(more fixes coming tomorrow)

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

[1]: https://github.com/reckenrode/nix-packages/tree/96b37cf5b91c092c924d4494d4d3fbb8adc48b65/pkgs/unit/ff/ffxiv

RaitoBezarius commented:
Hello, thank you for your contribution, please read the CONTRIBUTING guide and target staging adequately (without mass pinging everyone *by reading the guide*).

rrbutani commented on pkgs/stdenv/darwin/unpack-bootstrap-tools-aarch64.sh:
```diff
@@ -7,12 +7,22 @@ $bzip2 -d < $tarball | (cd $out && $cpio -i)
 
 export PATH=$out/bin
 
+# Work around codesign that does not understand --sign, which causes compiler-rt to fail to build.
+# This can be removed once bootstrapTools is updated with a newer codesign.
+mv $out/bin/codesign $out/bin/.codesign-wrapped
+
+cat <<CODESIGN > $out/bin/codesign
+#!/$out/bin/bash
+$out/bin/.codesign-wrapped "\${@/--sign/-s}"
+CODESIGN
+chmod a+x $out/bin/codesign
```
@reckenrode I think you mentioned (on Matrix) trying to build and use the newer `sigtool` instead of using `sigtool` from bootstrap; do you remember why that didn't work?

reckenrode replied:
The `compiler-rt` build still used `codesign` from the bootstrap tools even though I’d tried to override the `sigtool` package with the newer one. I didn’t dig into it too deeply because of how lengthy `stdenv` rebuilds are. I wanted to get a complete LLVM 15 `stdenv` built before going back to clean things up.

reckenrode replied:
After digging into it some more, what appears to be happening is `codesign` is leaking from the bootstrap tools via derivations (such as `coreutils` or `clang-unwrapped`) that just link the tools `bin` to `$out/bin`, and that’s what the LLVM build is finding instead of the one that’s built in stage 1.

reckenrode replied:
I fixed the issue with codesign as part of the bigger refactor.

reckenrode committed:
pkg-config: fix build with newer clang

reckenrode committed:
darwin.stdenv: refactor stdenv definition

In preparation for bumping the LLVM used by Darwin, this change
refactors and reworks the stdenv build process. When it made sense,
existing behaviors were kept to avoid causing any unwanted breakage.
However, there are some differences. The reasoning and differences are
discussed below.

- Improved cycle times - Working on the Darwin stdenv was a tedious
  process because `allowedRequisites` determined what was allowed
  between stages. If you made a mistake, you might have to wait a
  considerable amount of time for the build to fail. Using assertions
  makes many errors fail at evaluation time and makes moving things
  around safer and easier to do.
- Decoupling from bootstrap tools - The stdenv build process builds as
  much as it can in the early stages to remove the requirement that the
  bootstrap tools need bumped in order to bump the stdenv itself. This
  should lower the barrier to updates and make it easier to bump in the
  future. It also allows changes to be made without requiring additional
  tools be added to the bootstrap tools.
- Patterned after the Linux stdenv - I tried to follow the patterns
  established in the Linux stdenv with adaptations made to Darwin’s
  needs. My hope is this makes the Darwin stdenv more approable for
  non-Darwin developers who made need to interact with it. It also
  allowed some of the hacks to be removed.
- Documentation - Comments were added explaining what was happening and
  why things were being done. This is particular important for some
  stages that might not be obvious (such as the sysctl stage).
- Cleanup - Converting the intermediate `allowedRequisites` to
  assertions revealed that many packages were being referenced that no
  longer exist or have been renamed. Removing them reduces clutter and
  should help make the stdenv bootstrap process be more understandable.

reckenrode committed:
indent: fix build with clang 13 or newer

reckenrode committed:
rewrite-tbd: avoid infinite recursion when cmake is not cmakeMinimal

eliasnaur commented on pkgs/stdenv/darwin/default.nix:
```diff
@@ -314,403 +308,604 @@ rec {
+      preHook = commonPreHook + ''
+          export PATH_LOCALE=${prevStage.darwin.locale}/share/locale
+        ''
+        # Having this set when trying to cross-compile to Linux causes problems, so only set it
+        # when building for Darwin.
```
Nit: consider dropping comment. It seems clear that darwin specific stuff should be guarded by platform checks, by principle.

eliasnaur commented on pkgs/stdenv/darwin/default.nix:
```diff
@@ -314,403 +308,604 @@ rec {
-      initialPath = import ../generic/common-path.nix { inherit pkgs; };
-      shell = "${pkgs.bash}/bin/bash";
+      preHook = commonPreHook + ''
+          export PATH_LOCALE=${prevStage.darwin.locale}/share/locale
```
Why not guard `PATH_LOCALE` by the `targetPlatform` check as well?

reckenrode replied:
It didn’t occur to me at the time, but I’ve fixed it in my local branch. I’ll be pushing an update to move `PATH_LOCALE` under the `targetPlatform` check and removing the comment later today (hopefully after I resolve the issue with an infinite recursion building xnu on x86_64-darwin).

reckenrode commented on lib/systems/platforms.nix:
```diff
@@ -385,8 +385,8 @@ rec {
     gcc = {
-      arch = "armv8.3-a+crypto+sha2+aes+crc+fp16+lse+simd+ras+rdm+rcpc";
-      cpu = "apple-a13";
+      arch = "armv8.5a+crc+crypto+dotprod+fp16fml+ras+lse+rdm+rcpc+sm4+sha3+sha2+aes";
+      cpu = "apple-m1";
```
Not as far as I can tell. There’s an open issue at iains/gcc-darwin-arm64#86 for support.

reckenrode committed:
xnu: fix build with newer clang

reckenrode committed:
cctools: make cctools-llvm the default on Darwin

reckenrode committed:
cctools-llvm: add at 15.0.7

cctools-llvm is a replacement for cctools that replaces as much of
cctools with equivalents from LLVM that it can reasonably do. This was
motivated by wanting to reduce dependencies on cctools, which are
updated infrequently by upstream.

To provide a motivating example, the version of `strip` included in
cctools cannot properly strip the archives in compiler-rt in LLVM 15.
Paths are left to bootstrap tools, resulting in failed requisites checks
in the final stdenv build. Since `strip` needs replaced, other tools
were replaced as well.

Note that this has to be done in cctools (or some equivalent) because
some derivations (noteably LLVM) use the bintools of the stdenv directly
instead of going through the wrapper.

The following tools from LLVM are not used in this derivation:

* LLD - not fully compatible with ld64 yet and potentially too big of a change;
* libtool - not a drop-in replacement yet because it does not support
  linker passthrough, which is needed by xcbuild; and
* install_name_tool - fails when trying to build swift-corefoundation.

If other incompatibilities are found, other tools can be reverted.
However, testing for the stdenv bump has shown the others seem to work
pretty well as replacements.

One final caveat: some tools are not duplicated or linked from cctools.
As a rule of thumb, the actual tools on a Darwin system were used to
determine which ones to link, so there are some differences (e.g., in
names or availability).

eliasnaur commented:
Gentle ping. Did you get further in fixing the recursion problem?

reckenrode commented:
> Gentle ping. Did you get further in fixing the recursion problem?

Sorry about the lack of updates here. I’ve been posting daily updates on Matrix.

I resolved the infinite several days ago and have been working on cleaning up the paths in the final output. There was a bug in cctools’s `strip` that was leaving references to the bootstrap tools paths in one of compiler-rt’s archives. The cycle time on fixing that was pretty awful (having to do a full build of the stdenv takes hours), so it took a few days to find and implement a solution.

The current status is x86_64-darwin built a clean stdenv as of this morning. I found a problem with xpc and configd when trying to build something with it, but I think I resolved it. I started another rebuilding of the stdenv, and I’m also building Wine as a test case.

I’m using Wine as a test case because its build process pulls in quite a bit of stuff (including building cross-compilers for the Windows DLLs). I might also try building some Haskell and Qt stuff.

If that all works, I’ll clean up the PR and the  mark it ready. I do want to move the CF hook into CF itself and apply it unconditionally, which I haven’t yet. The stdenv itself doesn’t use CF, so that should be okay to do. CF is only included as an `extraBuildInput`.

eliasnaur commented:
> > Gentle ping. Did you get further in fixing the recursion problem?
> 
> Sorry about the lack of updates here. I’ve been posting daily updates on Matrix.
> 

Thank you for the update!

> If that all works, I’ll clean up the PR and the mark it ready. I do want to move the CF hook into CF itself and apply it unconditionally, which I haven’t yet. The stdenv itself doesn’t use CF, so that should be okay to do. CF is only included as an `extraBuildInput`.

Moving the environment variables to CF sounds like what https://github.com/NixOS/nixpkgs/pull/230870 does. I mention it in case you need similar hacks to disable the hook during stdenv building.

reckenrode commented:
> Moving the environment variables to CF sounds like what #230870 does. I mention it in case you need similar hacks to disable the hook during stdenv building.

The original issue indicated it was done to avoid unwanted references between stages. I dropped `allowedRequisites` in favor of assertions in the reworked bootstrap. `allowedRequisites` and `disallowedRequisites` are now used only in the final stdenv, so that shouldn’t be an issue. This approach follows how the Linux stdenv is built.

reckenrode committed:
swift-corelibs: set NIX_COREFOUNDATION_RPATH in a hook

Closes #230870. Thanks to @eliasnaur for the test case and for rasining
awareness and to @veprbl for the work done on #111385.

This takes a slightly different approach from those two PRs. The hook is
set unconditionally. The stdenv bootstrap doesn’t really need CF at all,
so setting the hook is harmless. This simplifies things.

reckenrode committed:
configd: fix build with newer LLVM and bootstrap

Clang 15 does not like the fake xpc headers. Use the real ones instead.
Doing this no longer causes an infinite recursion because xnu now
depends on python3Minimal, which does not include configd support.

reckenrode committed:
xnu: use python3Minimal instead of python3

xnu does not require a Python that supports configd, so use a minimal
one without support. Doing this prevents an infinite recuresion when
building configd with xpc, which is needed to build with a newer LLVM.

reckenrode commented:
I pushed the current branch. I expect I’ll be opening separate PRs to fix things that don’t build with the new clang that aren’t really needed for the bootstrap (or fail immediately). I still need to do some cleanup on commit messages before marking this ready and review any other issues that could be resolved/affected. I dropped the `platforms.nix` changes for now.

reckenrode committed:
apple_sdk_11_0.stdenv: replace extraBuildInputs with CF from the 11.0 SDK

reckenrode committed:
darwin.stdenv: bump LLVM version to 15

reckenrode commented:
The last push should reflect the final PR. I am keeping this draft until I can get the PRs submitted for the failures due to the stdenv change, which should be tomorrow.

reckenrode commented:
The ofborg eval is failing on llvmPackages.clang-polly-unwrapped, but that attribute is not part of llvmPackages_15. It appears it’s not included been part of any llvmPackages since 11.

cole-h commented:
Running the outpaths.nix check locally, this is what I see right before the eval fails:

```
[...snip...]
instantiated 'llvm-manpages-15.0.7' -> '/nix/store/2msilg6yh3p7493r1xv7mdb0j432s4mz-llvm-manpages-15.0.7.drv'
evaluating attribute 'recurseForDerivations'
evaluating attribute 'llvmPackages'
evaluating attribute '__unfix__'
evaluating attribute 'recurseForDerivations'
evaluating attribute 'bintools'
evaluating file '/home/vin/workspace/vcs/nixpkgs/master/pkgs/development/compilers/llvm/11/bintools/default.nix'
evaluating attribute 'recurseForDerivations'
evaluating attribute 'bintools-unwrapped'
evaluating attribute 'recurseForDerivations'
evaluating attribute 'bintoolsNoLibc'
evaluating attribute 'recurseForDerivations'
evaluating attribute 'clang'
evaluating attribute 'aarch64-darwin'
evaluating attribute 'recurseForDerivations'
evaluating attribute 'clang-manpages'
evaluating attribute 'aarch64-darwin'
evaluating attribute 'recurseForDerivations'
evaluating attribute 'clang-polly-unwrapped'
evaluating attribute 'aarch64-darwin'
error:
[...snip...]
```


Note that this is specifically a problem on the aarch64-darwin eval -- all other platforms are fine. You can reproduce this by looking at https://github.com/NixOS/ofborg#running-meta-checks-locally and commenting out all the architectures except for aarch64-darwin.

Dunno what's going on there, but maybe this will help you diagnose.

reckenrode commented:
Closing this in favor of tracking issue #234710, so I can submit changes incrementally while working through the ofborg issue and the gnugrep issue on x86_64-darwin.
